File: src/main/java/de/officeryoda/fhysics/engine/Vector2Int.java
package de.officeryoda.fhysics.engine;

import lombok.Data;

/**
 * A class representing a 2D vector with int precision.
 * This class provides basic operations for vector manipulation.
 */
@Data
public class Vector2Int {

    /**
     * The x-component of the vector.
     */
    private int x;
    /**
     * The y-component of the vector.
     */
    private int y;

    /**
     * Constructs a new Vector2Int with specified x and y components.
     *
     * @param x The x-component of the vector.
     * @param y The y-component of the vector.
     */
    public Vector2Int(int x, int y) {
        this.x = x;
        this.y = y;
    }

    /**
     * Constructs a Vector2Int with default values (0, 0).
     */
    public Vector2Int() {
        this(0, 0);
    }

    /**
     * Gets a constant Vector2Int with components (0, 0).
     *
     * @return A Vector2Int with components (0, 0).
     */
    public static Vector2Int zero() {
        return new Vector2Int(0, 0);
    }

    /**
     * Sets the x and y components of the vector.
     *
     * @param x The new x-component.
     * @param y The new y-component.
     */
    public void set(int x, int y) {
        this.x = x;
        this.y = y;
    }

    /**
     * Sets the components of the vector based on another Vector2Int.
     *
     * @param vec2 The Vector2Int to copy components from.
     */
    public void set(Vector2Int vec2) {
        this.x = vec2.x;
        this.y = vec2.y;
    }

    /**
     * Adds another Vector2Int to this vector.
     *
     * @param other The Vector2Int to be added.
     * @return The updated Vector2Int after addition.
     */
    public Vector2Int add(Vector2Int other) {
        this.x += other.x;
        this.y += other.y;
        return this;
    }

    /**
     * Creates a new Vector2Int by adding another Vector2Int to this vector.
     *
     * @param other The Vector2Int to be added.
     * @return A new Vector2Int representing the sum of the two vectors.
     */
    public Vector2Int addNew(Vector2Int other) {
        return new Vector2Int(this.x + other.x, this.y + other.y);
    }

    /**
     * Subtracts another Vector2Int from this vector.
     *
     * @param other The Vector2Int to be subtracted.
     * @return The updated Vector2Int after subtraction.
     */
    public Vector2Int subtract(Vector2Int other) {
        this.x -= other.x;
        this.y -= other.y;
        return this;
    }

    /**
     * Creates a new Vector2Int by subtracting another Vector2Int from this vector.
     *
     * @param other The Vector2Int to be subtracted.
     * @return A new Vector2Int representing the difference of the two vectors.
     */
    public Vector2Int subtractNew(Vector2Int other) {
        return new Vector2Int(this.x - other.x, this.y - other.y);
    }

    /**
     * Multiplies the vector by a scalar.
     *
     * @param scalar The scalar value for multiplication.
     * @return The updated Vector2Int after multiplication.
     */
    public Vector2Int multiply(int scalar) {
        this.x *= scalar;
        this.y *= scalar;
        return this;
    }

    /**
     * Creates a new Vector2Int by multiplying the vector by a scalar.
     *
     * @param scalar The scalar value for multiplication.
     * @return A new Vector2Int representing the result of the multiplication.
     */
    public Vector2Int multiplyNew(int scalar) {
        return new Vector2Int(this.x * scalar, this.y * scalar);
    }

    /**
     * Calculates the magnitude (length) of the vector.
     *
     * @return The magnitude of the vector.
     */
    public double magnitude() {
        return Math.sqrt(sqrMagnitude());
    }

    /**
     * Calculates the squared magnitude of the vector.
     *
     * @return The squared magnitude of the vector.
     */
    public int sqrMagnitude() {
        return x * x + y * y;
    }

    /**
     * Calculates the Euclidean distance between this vector and another Vector2.
     *
     * @param other The Vector2 to calculate the distance to.
     * @return The Euclidean distance between this vector and the specified Vector2.
     */
    public double distance(Vector2Int other) {
        return Math.sqrt(sqrDistance(other));
    }

    /**
     * Calculates the squared Euclidean distance between this vector and another Vector2.
     * This method is computationally less expensive than distance() as it avoids the square root operation.
     *
     * @param other The Vector2 to calculate the squared distance to.
     * @return The squared Euclidean distance between this vector and the specified Vector2.
     */
    public int sqrDistance(Vector2Int other) {
        int dx = this.x - other.x;
        int dy = this.y - other.y;
        return dx * dx + dy * dy;
    }
}

File: src/main/java/de/officeryoda/fhysics/engine/Vector2Int.kt
package de.officeryoda.fhysics.engine

import kotlin.math.sqrt

/**
 * A 2D vector class providing basic vector operations.
 *
 * @property x The x-coordinate of the Vector2Int.
 * @property y The y-coordinate of the Vector2Int.
 */
data class Vector2Int
@JvmOverloads constructor(
    var x: Int = 0, var y: Int = 0
) {
    /**
     * Calculates the dot product of this Vector2Int with another Vector2Int.
     *
     * @param other The other Vector2Int.
     * @return The dot product of the two Vector2Int instances.
     */
    fun dot(other: Vector2Int): Int {
        return this.x * other.x + this.y * other.y
    }

    /**
     * Calculates the magnitude (length) of the Vector2Int.
     *
     * @return The magnitude of the Vector2Int.
     */
    fun magnitude(): Float {
        return sqrt(sqrMagnitude().toFloat())
    }

    /**
     * Calculates the squared magnitude (length) of the Vector2Int.
     *
     * @return The squared magnitude of the Vector2Int.
     */
    fun sqrMagnitude(): Int {
        return x * x + y * y
    }

    /**
     * Calculates the Euclidean distance between this Vector2Int and another Vector2Int.
     *
     * @param other The other Vector2Int.
     * @return The distance between the two Vector2Int instances.
     */
    fun distance(other: Vector2Int): Float {
        return sqrt(sqrDistance(other).toFloat())
    }

    /**
     * Calculates the squared Euclidean distance between this Vector2Int and another Vector2Int.
     *
     * @param other The other Vector2Int.
     * @return The squared distance between the two Vector2Int instances.
     */
    fun sqrDistance(other: Vector2Int): Int {
        val dx: Int = this.x - other.x
        val dy: Int = this.y - other.y
        return dx * dx + dy * dy
    }

    /**
     * Sets the components of this Vector2Int to be equal to another Vector2.
     *
     * @param other The Vector2Int whose components will be copied.
     */
    fun set(other: Vector2Int) {
        this.x = other.x
        this.y = other.y
    }

    /**
     * Adds another Vector2Int to this Vector2Int.
     *
     * @param other The Vector2Int to add.
     * @return The result of the addition.
     */
    operator fun plus(other: Vector2Int): Vector2Int {
        return Vector2Int(this.x + other.x, this.y + other.y)
    }

    /**
     * Subtracts another Vector2Int from this Vector2Int.
     *
     * @param other The Vector2Int to subtract.
     * @return The result of the subtraction.
     */
    operator fun minus(other: Vector2Int): Vector2Int {
        return Vector2Int(this.x - other.x, this.y - other.y)
    }

    /**
     * Multiplies the Vector2Int by a scalar.
     *
     * @param scalar The scalar value.
     * @return The result of the multiplication.
     */
    operator fun times(scalar: Float): Vector2Int {
        return Vector2Int((this.x * scalar).toInt(), (this.y * scalar).toInt())
    }

    /**
     * Divides the Vector2Int by a scalar.
     *
     * @param scalar The scalar value.
     * @return The result of the division.
     * @throws IllegalArgumentException if division by zero is attempted.
     */
    operator fun div(scalar: Float): Vector2Int {
        return Vector2Int((this.x / scalar).toInt(), (this.y / scalar).toInt())
    }

    /**
     * Adds another Vector2Int to this Vector2Int in-place.
     *
     * @param other The Vector2Int to add.
     */
    operator fun plusAssign(other: Vector2Int) {
        this.x += other.x
        this.y += other.y
    }

    /**
     * Subtracts another Vector2Int from this Vector2Int in-place.
     *
     * @param other The Vector2Int to subtract.
     */
    operator fun minusAssign(other: Vector2Int) {
        this.x -= other.x
        this.y -= other.y
    }

    /**
     * Multiplies the Vector2Int by a scalar in-place.
     *
     * @param scalar The scalar value.
     */
    operator fun timesAssign(scalar: Float) {
        this.x = this.x.times(scalar).toInt()
        this.y = this.y.times(scalar).toInt()
    }

    /**
     * Divides the Vector2Int by a scalar in-place.
     *
     * @param scalar The scalar value.
     * @throws IllegalArgumentException if division by zero is attempted.
     */
    operator fun divAssign(scalar: Float) {
        this.x = this.x.div(scalar).toInt()
        this.y = this.y.div(scalar).toInt()
    }

    /**
     * Returns a new Vector2Int instance that represents the negation of the current vector.
     *
     * @return The negated Vector2Int.
     */
    operator fun unaryMinus(): Vector2Int {
        return Vector2Int(-this.x, -this.y)
    }

    /**
     * Converts this Vector2Int to a Vector2 by converting the x and y coordinates to Float values.
     *
     * @return The converted Vector2.
     */
    fun toFloatVector2(): Vector2 {
        return Vector2(x.toFloat(), y.toFloat())
    }

    /**
     * Creates and returns a copy (clone) of the Vector2Int.
     *
     * @return A new Vector2Int object with the same x and y components.
     */
    fun clone(): Vector2Int {
        return Vector2Int(this.x, this.y)
    }

    /**
     * Returns a string representation of the Vector2Int.
     *
     * @return A string containing the values of the x and y components.
     */
    override fun toString(): String {
        return "Vector2Int(x=$x, y=$y)"
    }

    /**
     * Companion object providing utility functions for Vector2Int.
     */
    companion object {
        /**
         * Creates a Vector2Int with both components set to zero.
         *
         * @return The zero Vector2.
         */
        @JvmStatic
        val ZERO: Vector2Int
            get() = Vector2Int(0, 0)
    }
}
